testutil/integration/router.go

```diff
@@ -0,0 +1,97 @@
+package integration
+
+import (
+	"fmt"
+
+	"github.com/cometbft/cometbft/abci/types"
+	cmtproto "github.com/cometbft/cometbft/proto/tendermint/types"
+
+	"cosmossdk.io/log"
+	storetypes "cosmossdk.io/store/types"
+	dbm "github.com/cosmos/cosmos-db"
+
+	"github.com/cosmos/cosmos-sdk/baseapp"
+	"github.com/cosmos/cosmos-sdk/codec"
+	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/module"
+	authtx "github.com/cosmos/cosmos-sdk/x/auth/tx"
+)
+
+type App struct {
+	*baseapp.BaseApp
+
+	ctx    sdk.Context
+	logger log.Logger
+
+	queryHelper *baseapp.QueryServiceTestHelper
+}
+
+func NewIntegrationApp(name string, logger log.Logger, keys map[string]*storetypes.KVStoreKey, modules ...module.AppModuleBasic) *App {
+	db := dbm.NewMemDB()
+
+	interfaceRegistry := codectypes.NewInterfaceRegistry()
+	for _, module := range modules {
+		module.RegisterInterfaces(interfaceRegistry)
+	}
+
+	txConfig := authtx.NewTxConfig(codec.NewProtoCodec(interfaceRegistry), authtx.DefaultSignModes)
+
+	bApp := baseapp.NewBaseApp(fmt.Sprintf("integration-app-%s", name), logger, db, txConfig.TxDecoder())
+	bApp.MountKVStores(keys)
+	bApp.SetInitChainer(func(ctx sdk.Context, req types.RequestInitChain) (types.ResponseInitChain, error) {
+		return types.ResponseInitChain{}, nil
+	})
+
+	router := baseapp.NewMsgServiceRouter()
+	router.SetInterfaceRegistry(interfaceRegistry)
+	bApp.SetMsgServiceRouter(router)
+
+	if err := bApp.LoadLatestVersion(); err != nil {
+		panic(err)
```

Comment:
Consider wrapping this error to include more detail so the panic message is more immediately clear.